app/services/host_counter.rb

```diff
@@ -0,0 +1,28 @@
+class HostCounter
+  def initialize(association)
+    @association = association
+  end
+
+  def [](association)
+    counted_hosts[association.id] || 0
```

Comment:
perhaps use `&.id` in case for some reason a nil object is passed to this function?